python/ngen_cal/src/ngen/cal/_plugin_system.py

```diff
@@ -27,3 +27,17 @@ def setup_plugin_manager(plugins: list[Callable | ModuleType]) -> PluginManager:
             assert_never(plugin)
 
     return pm
+
+def setup_scoped_plugin_manager(spec: type, plugins: list[Callable | ModuleType]) -> PluginManager:
+    pm = pluggy.PluginManager(PROJECT_SLUG)
+    pm.add_hookspecs(spec)
+
+    for plugin in plugins:
+        if isinstance(plugin, Callable):
+            pm.register(plugin())
+        elif isinstance(plugin, ModuleType):
```

Comment:
If a module based plugin is registered, we will run into the same pickling a module issue in a multiprocessing context. I think we should just `assert not isinstance(plugin, ModuleType), "function plugins"` and catch that at the call site and re-raise with extra context.

Reply:
refactored in a4710a941b0869d1bf1c2d2bfd920b93b09646fc